### ISSUE TITLE
'NoneType' object has no attribute 'pk' in /user/<user>/stats/activity

### ISSUE THREAD

**Issue by jleclanche** [translate/pootle#3651]
See: http://pootle.lxde.org/user/jleclanche/stats/activity?month=2015-03&username=jleclanche

```pytb
Traceback:
File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  111.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/views/generic/base.py" in view
  69.             return self.dispatch(request, *args, **kwargs)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapper
  29.             return bound_func(*args, **kwargs)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/Pootle-2.7.0a1-py2.7.egg/pootle/apps/pootle_misc/util.py" in wrapper
  107.         return f(request, *args, **kwargs)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in bound_func
  25.                 return func.__get__(self, type(self))(*args2, **kwargs2)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/Pootle-2.7.0a1-py2.7.egg/pootle/apps/evernote_reports/views.py" in dispatch
  99.         return super(UserActivityView, self).dispatch(request, *args, **kwargs)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/Pootle-2.7.0a1-py2.7.egg/pootle/core/views.py" in dispatch
  91.             .dispatch(request, *args, **kwargs)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/views/generic/base.py" in dispatch
  87.         return handler(request, *args, **kwargs)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/Pootle-2.7.0a1-py2.7.egg/pootle/apps/evernote_reports/views.py" in get
  102.         data = get_activity_data(self.request, self.get_object(), self.month)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/Pootle-2.7.0a1-py2.7.egg/pootle/apps/evernote_reports/views.py" in get_activity_data
  430.         json['grouped'] = get_grouped_paid_words(scores, user, month)
File "/var/www/pootle/env/local/lib/python2.7/site-packages/Pootle-2.7.0a1-py2.7.egg/pootle/apps/evernote_reports/views.py" in get_grouped_paid_words
  572.         translated_words, reviewed_words = score.get_paid_words()
File "/var/www/pootle/env/local/lib/python2.7/site-packages/Pootle-2.7.0a1-py2.7.egg/pootle/apps/pootle_statistics/models.py" in get_paid_words
  615.         }.get(self.action_code, lambda: (None, None))()
File "/var/www/pootle/env/local/lib/python2.7/site-packages/Pootle-2.7.0a1-py2.7.egg/pootle/apps/pootle_statistics/models.py" in get_sugg_reviewed_accepted
  584.             suggester = self.submission.suggestion.user.pk

Exception Type: AttributeError at /user/jleclanche/stats/activity
Exception Value: 'NoneType' object has no attribute 'pk'
```

**Comment by julen** [comment]
We've never experienced this traceback so my guess is there's some incomplete migration in 2.6.

**Comment by jleclanche** [comment]
It rings a bit of a bell, it's possible it was in before 2.7 and that it's not a migration's fault. Do you know what causes a suggestion not to have a user? is that a normal thing?

**Comment by julen** [comment]
No it's not normal currently, but it could probably happen in the past, where there were two different models to track suggestions (actual suggestions and stats).

I notice the migrations that do the actual data move to the "new" model are different ([from our fork](https://github.com/translate/pootle/blob/101049d7904e53edaf0f2d313d9ff0be32912174/pootle/apps/pootle_store/migrations/0008_move_data_from_pootle_app_suggestion.py), [stable/2.6.0](https://github.com/translate/pootle/blob/stable/2.6.0/pootle/apps/pootle_store/migrations/0009_move_data_from_pootle_app_suggestion.py)). @ta2-1 might be able to shed some extra light here.

**Comment by ta2-1** [comment]
It's hard to say more without seeing your data before and after migration. 

```
mysql> select * from pootle_store_suggestion where user_id is null;
Empty set (0,04 sec)
```

`user_id` field doesn't have `NOT NULL` constraint but it should have actually.

```
mysql> desc pootle_store_suggestion;
+----------------------+-------------+------+-----+---------+----------------+
| Field                | Type        | Null | Key | Default | Extra          |
+----------------------+-------------+------+-----+---------+----------------+
| id                   | int(11)     | NO   | PRI | NULL    | auto_increment |
| target_f             | longtext    | NO   |     | NULL    |                |
| target_hash          | varchar(32) | NO   | MUL | NULL    |                |
| unit_id              | int(11)     | NO   | MUL | NULL    |                |
| user_id              | int(11)     | YES  | MUL | NULL    |                |
| translator_comment_f | longtext    | YES  |     | NULL    |                |
| reviewer_id          | int(11)     | YES  | MUL | NULL    |                |
| state                | varchar(16) | NO   | MUL | NULL    |                |
| creation_time        | datetime    | YES  | MUL | NULL    |                |
| review_time          | datetime    | YES  | MUL | NULL    |                |
+----------------------+-------------+------+-----+---------+----------------+
```
Let's add this.

**Comment by jleclanche** [comment]
Can we not add this in 2.7 please? Leave it for after the release, please.

**Comment by ta2-1** [comment]
No problem. We'll add this later.

**Comment by dwaynebailey** [comment]
On LO we have 2781 suggestions with a user_id of null.  We should be migrating those suggestions to 'nobody' or deleting them in the 2.6 migrations.

**Comment by unho** [comment]
Just proposed a migration that assigns orphan suggestions to `nobody`.

> `user_id` field doesn't have `NOT NULL` constraint but it should have actually.

Yes, perhaps it is time to do that change.

Also it can be a good idea to assign the suggestions for a user to `nobody` before deleting the user.

**Comment by unho** [comment]
A fix for this has been merged in 432751dc in stable/2.6.0 branch. So closing this.
